src/coreclr/jit/objectalloc.cpp

```diff
@@ -1334,3 +1482,1724 @@ void ObjectAllocator::RewriteUses()
+                return true;
+            }
+            JITDUMP("... guarded by failing GDV\n");
+            continue;
+        }
+        else
+        {
+            bool const isReachableOnGDVFailure = (trueSuccessorDominates && guardingRelop->OperIs(GT_NE)) ||
+                                                 (falseSuccessorDominates && guardingRelop->OperIs(GT_EQ));
+            if (isReachableOnGDVFailure)
+            {
+                info->m_block = idomBlock;
+                return true;
+            }
+            JITDUMP("... guarded by successful GDV\n");
+            continue;
+        }
+    }
+
+    JITDUMP("... no more doms\n");
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// IsGuard: does block look like a GDV guard
+//
+// Arguments:
+//   block -- block in question
+//   info -- [out] guard info
+//
+// Returns:
+//   Comparison tree if this is a guard, or nullptr
+//
+GenTree* ObjectAllocator::IsGuard(BasicBlock* block, GuardInfo* info)
+{
+    if (!block->KindIs(BBJ_COND))
+    {
+        JITDUMP("... not BBJ_COND\n");
+        return nullptr;
+    }
+
+    Statement* const stmt = block->lastStmt();
+    if (stmt == nullptr)
+    {
+        JITDUMP("... no stmt\n");
+        return nullptr;
+    }
+
+    GenTree* const jumpTree = stmt->GetRootNode();
+    if (!jumpTree->OperIs(GT_JTRUE))
+    {
+        JITDUMP("... no JTRUE\n");
+        return nullptr;
+    }
+
+    GenTree* const tree = jumpTree->AsOp()->gtOp1;
+
+    // Must be an equality or inequality
+    //
+    if (!tree->OperIs(GT_NE, GT_EQ))
+    {
+        JITDUMP("... not NE/EQ\n");
+        return nullptr;
+    }
+
+    GenTree* op1     = tree->AsOp()->gtOp1;
+    GenTree* op2     = tree->AsOp()->gtOp2;
+    bool     swapped = false;
+
+    // gdv creates NE(hnd, indir(locl))
+    // but let's not rely on that
+    //
+    if (!op1->OperIs(GT_IND))
+    {
+        swapped = true;
+        std::swap(op1, op2);
+    }
+
+    if (!op1->OperIs(GT_IND))
+    {
+        JITDUMP("... no JTRUE(cmp(ind, ...))\n");
+        return nullptr;
+    }
+
+    if (!op1->TypeIs(TYP_I_IMPL))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int, ...))\n");
+        return nullptr;
+    }
+
+    GenTree* const addr = op1->AsIndir()->Addr();
+
+    if (!addr->TypeIs(TYP_REF))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(*:ref), ...))\n");
+        return nullptr;
+    }
+
+    if (!addr->OperIs(GT_LCL_VAR))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(lcl:ref), ...))\n");
+        return nullptr;
+    }
+
+    if (!op2->IsIconHandle(GTF_ICON_CLASS_HDL))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(lcl:ref), clsHnd))\n");
+        return nullptr;
+    }
+
+    // Passed the checks... fill in the info.
+    //
+    info->m_local  = addr->AsLclVar()->GetLclNum();
+    bool isNonNull = false;
+    bool isExact   = false;
+    info->m_type   = (CORINFO_CLASS_HANDLE)op2->AsIntCon()->gtIconVal;
```

Comment:
Should probably use `gtGetHelperArgClassHandle` instead, afair your code won't work for NAOT/R2R as it needs compile-time handle

Reply:
There's no helper call here? This type is only used to verify other "equivalent" GDV tests, so it is ok to just use the raw icon value.

Reply:
> There's no helper call here?

the name of that function is a bit unfortunate. I guess if you only need to compare it against something it will work, but maybe change it to ssize_t? because presumably it's an invalid CORINFO_CLASS_HANDLE on AOT, any attempt to e.g. do `eeGetClassName` on it will crash

Reply:
Agree that this would be more appropriate as a `target_ssize_t`.

Is there a chance for this optimization in R2R? I guess the tree we would see here would be some kind of indirection for the handle, but in any case lots of the necessary inlining probably would be blocked anyway, so understandable to not bother.